satpy/readers/mirs.py

```diff
@@ -334,70 +327,122 @@ def _nan_for_dtype(data_arr_dtype):
+
+        # let file metadata take precedence over ds_info from yaml,
+        # but if yaml has more to offer, include it here, but fix
+        # units.
+        ds_info.update(data.attrs)
+
+        # special cases
+        if ds_info['name'] in ["latitude", "longitude"]:
+            ds_info["standard_name"] = ds_info.get("standard_name",
+                                                   ds_info['name'])
+
+        # try to assign appropriate units (if "Kelvin" covert to K)
+        units_convert = {"Kelvin": "K"}
+        data_unit = ds_info.get('units', None)
+        ds_info['units'] = units_convert.get(data_unit, data_unit)
+
+        scale = ds_info.pop('scale_factor', 1.0)
+        offset = ds_info.pop('add_offset', 0.)
+        fill_value = ds_info.pop("_FillValue", global_attr_fill)
+        valid_range = ds_info.pop('valid_range', None)
+
+        data = self._scale_data(data, scale, offset)
+        data = self._fill_data(data, fill_value, scale, offset)
+        data = self._apply_valid_range(data, valid_range, scale, offset)
+
+        data.attrs = ds_info
+
+        return data, ds_info
 
     def get_dataset(self, ds_id, ds_info):
         """Get datasets."""
         if 'dependencies' in ds_info.keys():
             idx = ds_info['channel_index']
             data = self['BT']
             data = data.rename(new_name_or_name_dict=ds_info["name"])
+            data, ds_info = self.apply_attributes(data, ds_info)
 
             if self.sensor.lower() == "atms" and self.limb_correction:
                 sfc_type_mask = self['Sfc_type']
                 data = limb_correct_atms_bt(data, sfc_type_mask,
                                             self._get_coeff_filenames,
                                             ds_info)
+
                 self.nc = self.nc.merge(data)
             else:
                 LOG.info("No Limb Correction applied.")
                 data = data[:, :, idx]
         else:
             data = self[ds_id['name']]
+            data, ds_info = self.apply_attributes(data, ds_info)
+
+        data.attrs = self.update_metadata(ds_info)
 
-        data.attrs = self.get_metadata(ds_info)
         return data
 
-    def _available_if_this_file_type(self, configured_datasets):
+    def available_datasets(self, configured_datasets=None):
+        """Dynamically discover what variables can be loaded from this file.
+
+        See :meth:`satpy.readers.file_handlers.BaseHandler.available_datasets`
+        for more information.
+
+        """
+        handled_vars = set()
         for is_avail, ds_info in (configured_datasets or []):
             if is_avail is not None:
                 # some other file handler said it has this dataset
                 # we don't know any more information than the previous
                 # file handler so let's yield early
                 yield is_avail, ds_info
                 continue
+
+            if self.file_type_matches(ds_info['file_type']):
```

Comment:
Is it possible to restructure this block of code so `file_type_matches` doesn't have to be called twice?